File: modules/uml/app/controllers/uml/UmlAdmin.java
package controllers.uml;

import javax.inject.Inject;

import com.fasterxml.jackson.databind.JsonNode;
import com.github.fge.jsonschema.core.report.ProcessingReport;

import controllers.core.AExerciseAdminController;
import model.StringConsts;
import model.UmlExTextParser;
import model.UmlExercise;
import model.UmlExerciseReader$;
import model.exercisereading.JsonReader$;
import model.user.User;
import play.data.DynamicForm;
import play.data.FormFactory;
import play.libs.Json;
import play.mvc.Result;
import play.twirl.api.Html;

public class UmlAdmin extends AExerciseAdminController<UmlExercise> {
  
  @Inject
  public UmlAdmin(FormFactory theFactory) {
    super(theFactory, UmlToolObject$.MODULE$, UmlExercise.finder, UmlExerciseReader$.MODULE$);
  }
  
  @Override
  public Result adminIndex() {
    return ok(views.html.umlAdmin.index.render(getUser()));
  }
  
  public Result checkSolution() {
    final DynamicForm form = factory.form().bindFromRequest();
    
    final JsonNode solNode = Json.parse(form.get(StringConsts.SOLUTION_NAME));
    final ProcessingReport report = JsonReader$.MODULE$.validateJson(solNode, UmlController.SOLUTION_SCHEMA_NODE).get();
    
    if(report.isSuccess())
      return ok("ok");
    
    return ok(report.toString());
  }
  
  public Result newExerciseStep2() {
    final UmlExercise exercise = exerciseReader.initFromForm(0, factory.form().bindFromRequest());
    final UmlExTextParser parser = new UmlExTextParser(exercise.text, exercise.mappings, exercise.ignoreWords);
    return ok(views.html.umlAdmin.newExerciseStep2Form.render(getUser(), exercise, parser.getCapitalizedWords()));
  }
  
  public Result newExerciseStep3() {
    final UmlExercise exercise = exerciseReader.initFromForm(0, factory.form().bindFromRequest());
    return ok(views.html.umlAdmin.newExerciseStep3Form.render(getUser(), exercise));
  }
  
  @Override
  protected Html renderExEditForm(User user, UmlExercise exercise, boolean isCreation) {
    // TODO Auto-generated method stub
    return null;
  }
  
}

File: modules/uml/app/controllers/uml/UmlAdmin.scala
package controllers.uml

import javax.inject.Inject

import controllers.excontrollers.AExerciseAdminController
import model.exercisereading.{JsonReader, ReadingError, ReadingFailure, ReadingResult}
import model.{UmlExTextParser, UmlExercise, UmlExerciseReader}
import play.api.mvc.ControllerComponents
import play.libs.Json

import scala.collection.JavaConverters._
import scala.util.{Failure, Success}

class UmlAdmin @Inject()(cc: ControllerComponents)
  extends AExerciseAdminController[UmlExercise](cc, UmlToolObject, UmlExercise.finder, UmlExerciseReader) {

  def checkSolution = Action { implicit request => {
    val solNode = Json.parse("" /* factory.form().bindFromRequest().get(StringConsts.SOLUTION_NAME)*/)
    JsonReader.validateJson(solNode, UmlController.SolutionSchemaNode) match {
      case Success(_) => Ok("Ok...")
      case Failure(_) => BadRequest("FEHLER!")
    }
  }
  }

  def newExerciseStep2 = Action { implicit request =>
//    exerciseReader.initFromForm(0, null /* factory.form().bindFromRequest()*/) match {
//      case ReadingError(_, _, _) => BadRequest("There has been an error...")
//      case ReadingFailure(_) => BadRequest("There has been an error...")
//      case ReadingResult(exercises) =>
//        val exercise = exercises.head.read.asInstanceOf[UmlExercise]
//        val parser = new UmlExTextParser(exercise.text, exercise.mappings.asScala.toMap, exercise.ignoreWords.asScala.toList)
//        Ok(views.html.umlAdmin.newExerciseStep2Form.render(getUser, exercise, parser.capitalizedWords.toList))
//    }
    Ok("TODO!")
  }

  def newExerciseStep3 = Action { implicit request =>
//    exerciseReader.initFromForm(0, null /* factory.form().bindFromRequest()*/) match {
    //      case ReadingError(_, _, _) => BadRequest("There has been an error...")
    //      case ReadingFailure(_) => BadRequest("There has been an error...")
    //      case ReadingResult(exercises) =>
    //        val exercise = exercises.head.read.asInstanceOf[UmlExercise]
    //        Ok(views.html.umlAdmin.newExerciseStep3Form.render(getUser, exercise))
    //    }
    Ok("TODO!")
  }

}
